src/transformation/visitors/binary-expression/compound.ts

```diff
@@ -85,74 +70,137 @@ export function transformCompoundAssignment(
         const accessExpression = lua.createTableIndexExpression(obj, index);
 
-        const tmp = lua.createIdentifier("____tmp");
+        const tmp = context.createTempNameForLuaExpression(left);
         let tmpDeclaration: lua.VariableDeclarationStatement;
         let assignStatement: lua.AssignmentStatement;
+        let operatorExpression: lua.Expression;
         if (isPostfix) {
             // local ____tmp = ____obj[____index];
             // ____obj[____index] = ____tmp ${replacementOperator} ${right};
             tmpDeclaration = lua.createVariableDeclarationStatement(tmp, accessExpression);
-            const operatorExpression = transformBinaryOperation(context, tmp, right, operator, expression);
+            [rightPrecedingStatements, operatorExpression] = transformBinaryOperation(
+                context,
+                tmp,
+                right,
+                rightPrecedingStatements,
+                operator,
+                expression
+            );
             assignStatement = lua.createAssignmentStatement(accessExpression, operatorExpression);
         } else {
             // local ____tmp = ____obj[____index] ${replacementOperator} ${right};
             // ____obj[____index] = ____tmp;
-            const operatorExpression = transformBinaryOperation(context, accessExpression, right, operator, expression);
+            [rightPrecedingStatements, operatorExpression] = transformBinaryOperation(
+                context,
+                accessExpression,
+                right,
+                rightPrecedingStatements,
+                operator,
+                expression
+            );
             tmpDeclaration = lua.createVariableDeclarationStatement(tmp, operatorExpression);
             assignStatement = lua.createAssignmentStatement(accessExpression, tmp);
         }
         // return ____tmp
-        return { statements: [objAndIndexDeclaration, tmpDeclaration, assignStatement], result: tmp };
+        return {
+            statements: [objAndIndexDeclaration, ...rightPrecedingStatements, tmpDeclaration, assignStatement],
+            result: tmp,
+        };
     } else if (isPostfix) {
         // Postfix expressions need to cache original value in temp
         // local ____tmp = ${left};
         // ${left} = ____tmp ${replacementOperator} ${right};
         // return ____tmp
-        const tmpIdentifier = lua.createIdentifier("____tmp");
+        const tmpIdentifier = context.createTempNameForLuaExpression(left);
         const tmpDeclaration = lua.createVariableDeclarationStatement(tmpIdentifier, left);
-        const operatorExpression = transformBinaryOperation(context, tmpIdentifier, right, operator, expression);
-        const assignStatements = transformAssignment(context, lhs, operatorExpression);
+        let operatorExpression: lua.Expression;
+        [rightPrecedingStatements, operatorExpression] = transformBinaryOperation(
```

Comment:
Personal opinion: I think reassigning existing variables here instead of just making consts in the scope where they are relevant is pretty ugly.